File: app/src/main/java/pl/laskowski/marcin/solving/deduction/algorithm/DeductionAlgorithm.java
package pl.laskowski.marcin.solving.deduction.algorithm;

import pl.laskowski.marcin.model.Region;
import pl.laskowski.marcin.model.SudokuHintGrid;

import java.util.List;

public abstract class DeductionAlgorithm {

    public interface Factory {
        DeductionAlgorithm instance(List<Region> regions, SudokuHintGrid possibilities);
    }

    final List<Region> regions;
    final SudokuHintGrid possibilities;
    private boolean haveChanged = false;

    DeductionAlgorithm(List<Region> regions, SudokuHintGrid possibilities) {
        this.regions = regions;
        this.possibilities = possibilities;
    }

    public boolean solve() {
        for (Region region : regions) {
            solve(region);
        }
        return haveChanged;
    }

    abstract void solve(Region region);

    void notifyChange() {
        haveChanged = true;
    }

    boolean haveChanged() {
        return haveChanged;
    }

}

File: core/src/main/kotlin/io/github/ilikeyourhat/kudoku/solving/deduction/algorithm/DeductionAlgorithm.kt
package io.github.ilikeyourhat.kudoku.solving.deduction.algorithm

import io.github.ilikeyourhat.kudoku.model.Region
import io.github.ilikeyourhat.kudoku.model.hint.SudokuHintGrid

abstract class DeductionAlgorithm(
    val regions: List<Region>,
    val possibilities: SudokuHintGrid
) {

    interface Factory {
        fun instance(regions: List<Region>, possibilities: SudokuHintGrid): DeductionAlgorithm
    }

    open fun solve(): Boolean {
        return regions.asSequence()
            .map { solve(it) }
            .find { it } ?: false
    }

    abstract fun solve(region: Region): Boolean
}
